pyqtorch/circuit.py

```diff
@@ -74,15 +75,29 @@ def expectation(
             )
 
     @property
-    def _device(self) -> torch.device:
-        try:
-            (_, buffer) = next(self.named_buffers())
-            return buffer.device
-        except StopIteration:
-            return torch.device("cpu")
-
-    def init_state(self, batch_size: int = 1) -> torch.Tensor:
+    def _device(self) -> device:
+        device_lst: list[device] = []
+        for op in self.operations:
+            if isinstance(op, QuantumCircuit):
+                device_lst.append(op._device)
+            elif isinstance(op, Module):
+                device_lst += [b.device for b in op.buffers()]
+        maybe_device = list(set(device_lst))
+        if len(maybe_device) > 1:
+            logger.error(f"Found more than one device in module {self}.")
+        elif len(maybe_device) == 0:
+            logger.warning(f"Unable to determine device of module {self}.")
+        else:
+            _device = device_lst[0]
+            logger.info(f"Found device {_device}.")
+            return _device
```

Comment:
How often would this be called?

I think maybe `logger.debug` would be more appropriate.


also, you do an error log but instead of raising you return None. I think that should either not be logged error, or raise an exception.

I'd suggest that instead you rewrite as
```python
if len(maybe_device) == 1:
      ...
 else:
        logger.warning(f"Unable to determine device of module {self}. Found {maybe_device}, expected exactly one device")
```

Reply:
same answer as below. once for determining the device on which the module buffers sit and then initializing  an initial state on the same device

Reply:
yes, that I understand. 
I'll rewrite for clarity

Reply:
I was on my phone and ended up with multiple questions in a single comment, sorry about that.

Reply:
rewritten to match your suggestions

Reply:
thanks!